### ISSUE TITLE
The usage of finetuning of training-aware pruner

### ISSUE THREAD

**Issue by hobbitlzy** [microsoft/nni#5570]
**Describe the issue**:
I am using the code from [new_pruning_bert_glue.py](https://github.com/microsoft/nni/blob/master/examples/tutorials/new_pruning_bert_glue.py). It prunes the attention, ffn, and embedding in three subsequent stages. For each stage, the pruning pipeline is 
1. Apply certain pruner, training-aware pruner in this example.
2. Compress the model physically by `Speedup`.
3. Finetune the compressed model by distillaiton.

I am wondering why the finetuning is placed after the `Speedup`. The confusion arises from the unclear of the function of the training-aware pruner and `Speedup`. According to my understanding, the `Speedup` doesn't change the computation results of the model but only transform the fake pruning to real pruning, so the finetuning serves for the pruning. However, I suppose the training-aware pruner directly produces a finished pruned model Why is it necessary for the finetuning? 

I am new to `nni` and pruning, please correct me if I have any misunderstanding. :)

**Environment**:
Not applied.

**Configuration**:
Not applied.

**Log message**:
Not applied.

**How to reproduce it?**:
Not applied.

**Comment by J-shang** [comment]
hello @hobbitlzy , it is a good question, and you are right, `Speedup` transforms the fake pruning to real pruning, but there are some differences:
1. `Speedup` is a structure pruning speedup, that means it replaces the original module to a smaller one by removing several input dim/output dim weights (input features or output features for `Linear`, input channels or output channels for `Conv`). And in attention pruning, we use block sparse to sparse the weight, but `Speedup` only remove the all-masked attn head, the head that partial-masked will be recovered, so we add a finetuning stage after pruning attention. (We are working on block sparse speedup and n: m sparse speedup : ) ).
2. For training-aware concerns, even though these training-aware algorithms often have higher precision compared with post-training algorithms, but that don't mean continue finetuning cannot improve the precision, according to our experience, finetuning after pruning helps to improve the final performance in most scenarios, and you can also call these finetuning stages retraining or cooldown.

**Comment by hobbitlzy** [comment]
Thanks for your detail reply. I am still a little confused about the first response. I notice the model is unwrapped before feeding into `Speedup`, and the following fintuning also uses the unwrapped model. Combined with your explanation, can I interpret the reason of finetuning as the reamined heads will be treated as dense and previously masked weights are finetuned again?

**Comment by J-shang** [comment]
1. Yes, for example, a 12 heads attention layer may reduce to 6 heads after `Speedup`, and finetuning take these 6 dense heads.
2. The choice of pruner hyperparameters has not been tried a lot. The training length of each epochs is set to the epochs required for bert finetuning on the mnli dataset. According to our experience, gradually increasing the sparsity helps to improve the final performance, so we roughly set a moderate `interval_steps` (1000) to ensure a larger `total_times` (36000 / 1000 = 36) during the training process of 3 epochs (about 36000 steps).
